[Test Framework] Integrate apps with Galley

nmittler opened the pull request:
The most recent refactoring broke the apps component when Pilot is being used with Galley. The apps register their services with the ServiceManager directly. When Pilot is configured with Galley, however, it doesn't use the ServiceManager, which means that the app services are never properly registered with Pilot.

- Changed the Pilot and Apps component to require Galley to be configured, to avoid confusion.

- Removed the ServiceManager altogether - Galley is used for service registration.

Fixes #13090

istio-testing commented:
[APPROVALNOTIFIER] This PR is **NOT APPROVED**

This pull-request has been approved by: *<a href="https://github.com/istio/istio/pull/13115#" title="Author self-approved">nmittler</a>*
To fully approve this pull request, please assign additional approvers.
We suggest the following additional approver: **ozevren**

If they are not already assigned, you can assign the PR to them by writing `/assign @ozevren` in a comment when ready.

The full list of commands accepted by this bot can be found [here](https://go.k8s.io/bot-commands).

The pull request process is described [here](https://git.k8s.io/community/contributors/guide/owners.md#the-code-review-process)

<details open>
Needs approval from an approver in each of these files:

- **[galley/OWNERS](https://github.com/istio/istio/blob/master/galley/OWNERS)**
- ~~[pkg/OWNERS](https://github.com/istio/istio/blob/master/pkg/OWNERS)~~ [nmittler]
- ~~[tests/integration/OWNERS](https://github.com/istio/istio/blob/master/tests/integration/OWNERS)~~ [nmittler]

Approvers can indicate their approval by writing `/approve` in a comment
Approvers can cancel approval by writing `/approve cancel` in a comment
</details>
<!-- META={"approvers":["ozevren"]} -->

istio-testing commented:
@nmittler: The following test **failed**, say `/retest` to rerun them all:

Test name | Commit | Details | Rerun command
--- | --- | --- | ---
prow/istio-pilot-multicluster-e2e.sh | 0eca595d8a0d7ff8261decebd195cee135d3cebe | [link](https://k8s-gubernator.appspot.com/build/istio-prow/pr-logs/pull/istio_istio/13115/istio-pilot-multicluster-e2e/10629) | `/test istio-pilot-multicluster-e2e`



<details>

Instructions for interacting with me using PR comments are available [here](https://git.k8s.io/community/contributors/guide/pull-requests.md).  If you have questions or suggestions related to my behavior, please file an issue against the [kubernetes/test-infra](https://github.com/kubernetes/test-infra/issues/new?title=Prow%20issue:) repository. I understand the commands that are listed [here](https://go.k8s.io/bot-commands).
</details>
<!-- test report -->

istio-testing commented:
@nmittler: The following tests **failed**, say `/retest` to rerun them all:

Test name | Commit | Details | Rerun command
--- | --- | --- | ---
prow/istio-pilot-multicluster-e2e.sh | 0eca595d8a0d7ff8261decebd195cee135d3cebe | [link](https://k8s-gubernator.appspot.com/build/istio-prow/pr-logs/pull/istio_istio/13115/istio-pilot-multicluster-e2e/10629) | `/test istio-pilot-multicluster-e2e`
prow/istio-integ-k8s-tests.sh | 0eca595d8a0d7ff8261decebd195cee135d3cebe | [link](https://k8s-gubernator.appspot.com/build/istio-prow/pr-logs/pull/istio_istio/13115/istio-integ-k8s-tests/6313) | `/test istio-integ-k8s-tests`



<details>

Instructions for interacting with me using PR comments are available [here](https://git.k8s.io/community/contributors/guide/pull-requests.md).  If you have questions or suggestions related to my behavior, please file an issue against the [kubernetes/test-infra](https://github.com/kubernetes/test-infra/issues/new?title=Prow%20issue:) repository. I understand the commands that are listed [here](https://go.k8s.io/bot-commands).
</details>
<!-- test report -->

istio-testing commented:
@nmittler: PR needs rebase.

<details>

Instructions for interacting with me using PR comments are available [here](https://git.k8s.io/community/contributors/guide/pull-requests.md).  If you have questions or suggestions related to my behavior, please file an issue against the [kubernetes/test-infra](https://github.com/kubernetes/test-infra/issues/new?title=Prow%20issue:) repository.
</details>

istio-testing commented:
[APPROVALNOTIFIER] This PR is **NOT APPROVED**

This pull-request has been approved by: *<a href="https://github.com/istio/istio/pull/13115#pullrequestreview-223896239" title="LGTM">howardjohn</a>*, *<a href="https://github.com/istio/istio/pull/13115#" title="Author self-approved">nmittler</a>*
To fully approve this pull request, please assign additional approvers.
We suggest the following additional approver: **ozevren**

If they are not already assigned, you can assign the PR to them by writing `/assign @ozevren` in a comment when ready.

The full list of commands accepted by this bot can be found [here](https://go.k8s.io/bot-commands).

The pull request process is described [here](https://git.k8s.io/community/contributors/guide/owners.md#the-code-review-process)

<details open>
Needs approval from an approver in each of these files:

- **[galley/OWNERS](https://github.com/istio/istio/blob/master/galley/OWNERS)**
- ~~[pkg/OWNERS](https://github.com/istio/istio/blob/master/pkg/OWNERS)~~ [nmittler]
- ~~[tests/integration/OWNERS](https://github.com/istio/istio/blob/master/tests/integration/OWNERS)~~ [nmittler]

Approvers can indicate their approval by writing `/approve` in a comment
Approvers can cancel approval by writing `/approve cancel` in a comment
</details>
<!-- META={"approvers":["ozevren"]} -->

istio-testing commented:
[APPROVALNOTIFIER] This PR is **APPROVED**

This pull-request has been approved by: *<a href="https://github.com/istio/istio/pull/13115#pullrequestreview-223896239" title="LGTM">howardjohn</a>*, *<a href="https://github.com/istio/istio/pull/13115#" title="Author self-approved">nmittler</a>*

The full list of commands accepted by this bot can be found [here](https://go.k8s.io/bot-commands).

The pull request process is described [here](https://git.k8s.io/community/contributors/guide/owners.md#the-code-review-process)

<details >
Needs approval from an approver in each of these files:

- ~~[pkg/OWNERS](https://github.com/istio/istio/blob/master/pkg/OWNERS)~~ [nmittler]
- ~~[tests/integration/OWNERS](https://github.com/istio/istio/blob/master/tests/integration/OWNERS)~~ [nmittler]

Approvers can indicate their approval by writing `/approve` in a comment
Approvers can cancel approval by writing `/approve cancel` in a comment
</details>
<!-- META={"approvers":[]} -->

nmittler commented:
@howardjohn @smawson PTAL

I've added a hold for now. I've just pushed a major change to refactor the echo component so that it can be a complete replacement for the apps component. I've added full capabilities for both native and kube.  I've tested native, but will work tomorrow on swapping over the existing apps tests to use echo instead.

istio-testing commented:
@nmittler: The following tests **failed**, say `/retest` to rerun them all:

Test name | Commit | Details | Rerun command
--- | --- | --- | ---
prow/istio-integ-k8s-tests.sh | 0eca595d8a0d7ff8261decebd195cee135d3cebe | [link](https://k8s-gubernator.appspot.com/build/istio-prow/pr-logs/pull/istio_istio/13115/istio-integ-k8s-tests/6313) | `/test istio-integ-k8s-tests`
prow/istio-pilot-multicluster-e2e.sh | 5dcf36e9feed3679a954a5414a3f559717578c90 | [link](https://k8s-gubernator.appspot.com/build/istio-prow/pr-logs/pull/istio_istio/13115/istio-pilot-multicluster-e2e/10722) | `/test istio-pilot-multicluster-e2e`



<details>

Instructions for interacting with me using PR comments are available [here](https://git.k8s.io/community/contributors/guide/pull-requests.md).  If you have questions or suggestions related to my behavior, please file an issue against the [kubernetes/test-infra](https://github.com/kubernetes/test-infra/issues/new?title=Prow%20issue:) repository. I understand the commands that are listed [here](https://go.k8s.io/bot-commands).
</details>
<!-- test report -->

istio-testing commented:
@nmittler: The following tests **failed**, say `/retest` to rerun them all:

Test name | Commit | Details | Rerun command
--- | --- | --- | ---
prow/istio-pilot-multicluster-e2e.sh | 5dcf36e9feed3679a954a5414a3f559717578c90 | [link](https://k8s-gubernator.appspot.com/build/istio-prow/pr-logs/pull/istio_istio/13115/istio-pilot-multicluster-e2e/10722) | `/test istio-pilot-multicluster-e2e`
prow/istio-integ-k8s-tests.sh | 5dcf36e9feed3679a954a5414a3f559717578c90 | [link](https://k8s-gubernator.appspot.com/build/istio-prow/pr-logs/pull/istio_istio/13115/istio-integ-k8s-tests/6418) | `/test istio-integ-k8s-tests`



<details>

Instructions for interacting with me using PR comments are available [here](https://git.k8s.io/community/contributors/guide/pull-requests.md).  If you have questions or suggestions related to my behavior, please file an issue against the [kubernetes/test-infra](https://github.com/kubernetes/test-infra/issues/new?title=Prow%20issue:) repository. I understand the commands that are listed [here](https://go.k8s.io/bot-commands).
</details>
<!-- test report -->

istio-testing commented:
@nmittler: The following tests **failed**, say `/retest` to rerun them all:

Test name | Commit | Details | Rerun command
--- | --- | --- | ---
prow/istio-integ-k8s-tests.sh | 5dcf36e9feed3679a954a5414a3f559717578c90 | [link](https://k8s-gubernator.appspot.com/build/istio-prow/pr-logs/pull/istio_istio/13115/istio-integ-k8s-tests/6418) | `/test istio-integ-k8s-tests`
prow/istio-integ-local-tests.sh | 4028704c6c49956f2e78897b520ba25b4e52e4cd | [link](https://k8s-gubernator.appspot.com/build/istio-prow/pr-logs/pull/istio_istio/13115/istio-integ-local-tests/8281) | `/test istio-integ-local-tests`
prow/istio-pilot-multicluster-e2e.sh | 4028704c6c49956f2e78897b520ba25b4e52e4cd | [link](https://k8s-gubernator.appspot.com/build/istio-prow/pr-logs/pull/istio_istio/13115/istio-pilot-multicluster-e2e/10745) | `/test istio-pilot-multicluster-e2e`



<details>

Instructions for interacting with me using PR comments are available [here](https://git.k8s.io/community/contributors/guide/pull-requests.md).  If you have questions or suggestions related to my behavior, please file an issue against the [kubernetes/test-infra](https://github.com/kubernetes/test-infra/issues/new?title=Prow%20issue:) repository. I understand the commands that are listed [here](https://go.k8s.io/bot-commands).
</details>
<!-- test report -->

istio-testing commented:
@nmittler: The following tests **failed**, say `/retest` to rerun them all:

Test name | Commit | Details | Rerun command
--- | --- | --- | ---
prow/istio-integ-local-tests.sh | 4028704c6c49956f2e78897b520ba25b4e52e4cd | [link](https://k8s-gubernator.appspot.com/build/istio-prow/pr-logs/pull/istio_istio/13115/istio-integ-local-tests/8281) | `/test istio-integ-local-tests`
prow/istio-pilot-multicluster-e2e.sh | 4028704c6c49956f2e78897b520ba25b4e52e4cd | [link](https://k8s-gubernator.appspot.com/build/istio-prow/pr-logs/pull/istio_istio/13115/istio-pilot-multicluster-e2e/10745) | `/test istio-pilot-multicluster-e2e`
prow/istio-integ-k8s-tests.sh | 4028704c6c49956f2e78897b520ba25b4e52e4cd | [link](https://k8s-gubernator.appspot.com/build/istio-prow/pr-logs/pull/istio_istio/13115/istio-integ-k8s-tests/6445) | `/test istio-integ-k8s-tests`



<details>

Instructions for interacting with me using PR comments are available [here](https://git.k8s.io/community/contributors/guide/pull-requests.md).  If you have questions or suggestions related to my behavior, please file an issue against the [kubernetes/test-infra](https://github.com/kubernetes/test-infra/issues/new?title=Prow%20issue:) repository. I understand the commands that are listed [here](https://go.k8s.io/bot-commands).
</details>
<!-- test report -->

nmittler commented:
@smawson 

> This PR is rather big, is it possible to split into more manageable pieces? Can we get here more incrementally?

Yes, sorry ... I've reverted the changes to echo and moved them over to #13175.

> - It feels like we're doing a lot of hackery around how we create the sidecars, I'd like us to figure out how to clean that up and have a roadmap for making this all much simpler. That may require work on our actual components to make them run the way we want, but we should do that anyway :)
> - I initially had planned on having a separate "sidecar" or "envoy" component rather than having apps take care of it, and using as much of the real code as possible. We should discuss how we want that to look in the long term at some point (I only have half-baked opinions but something doesn't quite feel right here).

Absolutely agree. This PR is really just taking baby steps to:

- Fix a bug in apps that was introduced with the new framework changes
- Merge 2 components that are really the same thing (apps and echo).

I've raised #13177 to track progress toward addressing the hackery. This is very important.

nmittler commented on pkg/test/framework/components/apps/agent/bootstrap.go:
```diff
@@ -0,0 +1,179 @@
+	envoyBootstrapTemplate = getEnvoyBootstrapTemplate()
+)
+
+func getEnvoyBootstrapTemplate() *template.Template {
```
Agreed. I've mentioned that in #13177

nmittler commented on pkg/test/framework/components/echo/native/discoveryfilter.go:
```diff
@@ -0,0 +1,372 @@
+			if err != nil {
+				return nil, err
+			}
+			if inbound {
```
Understood. Do you think this should be addressed now, or can it wait for the resolutoion of #13177?

nmittler commented:
/test istio-integ-local-tests

howardjohn commented on pkg/test/framework/components/apps/agent/serviceentry.go:
```diff
@@ -0,0 +1,83 @@
+	w := bufio.NewWriter(&filled)
+	if err := serviceEntryTemplate.Execute(w, map[string]interface{}{
+		"ServiceName": a.cfg.ServiceName,
+		"Version":     a.cfg.Version,
```
Why do we have version? Seems like we just set it as a label on the ServiceEntry, and we never have different versions of the same service so I don't see why it is useful

nmittler commented on tests/integration/security/healthcheck/mtls_healthcheck_test.go:
```diff
@@ -74,8 +74,8 @@ spec:
-	pilot := pilot.NewOrFail(t, ctx, pilot.Config{})
-	aps := apps.NewOrFail(ctx, t, apps.Config{Pilot: pilot, AppParams: []apps.AppParam{
+	p := pilot.NewOrFail(t, ctx, pilot.Config{})
+	aps := apps.NewOrFail(t, ctx, apps.Config{Pilot: p, AppParams: []apps.AppParam{
```
Both now call validate. The namespace should have been optional ... we now use a default namespace if not provided.

howardjohn commented on pkg/test/framework/components/apps/native_test.go:
```diff
@@ -0,0 +1,67 @@
+)
+
+func TestNative(t *testing.T) {
+	ctx := framework.NewContext(t)
```
Looking at the code, looks like it just does some extra logging and ensures you close the context - not a big difference. Doesn't seem like a big deal, we may just want to change all tests to the same in a later PR for consistency

howardjohn commented on pkg/test/framework/components/environment/native/native.go:
```diff
@@ -35,19 +44,29 @@ type Environment struct {
+		Mesh:            &mesh,
+		SystemNamespace: systemNamespace,
+		Domain:          domain,
+		PortManager:     portMgr,
```
Looks like portmanager in Environment?

howardjohn commented on pkg/test/framework/components/apps/native.go:
```diff
@@ -101,37 +106,46 @@ func newNative(ctx resource.Context, cfg Config) (Instance, error) {
+	}
 
-		app, err := newNativeApp(cfg)
+	portManager, err := reserveport.NewPortManager()
```
https://github.com/istio/istio/pull/13115/files#diff-2368df1f890905899c2a72cb05f0afb2R69? Not sure if that links right, but it seems to be in this one?

howardjohn replied:
If its cleaned up in the next one that is fine too

istio-testing commented:
[APPROVALNOTIFIER] This PR is **APPROVED**

This pull-request has been approved by: *<a href="https://github.com/istio/istio/pull/13115#pullrequestreview-225195380" title="LGTM">howardjohn</a>*, *<a href="https://github.com/istio/istio/pull/13115#" title="Author self-approved">nmittler</a>*

The full list of commands accepted by this bot can be found [here](https://go.k8s.io/bot-commands).

The pull request process is described [here](https://git.k8s.io/community/contributors/guide/owners.md#the-code-review-process)

<details >
Needs approval from an approver in each of these files:

- ~~[pkg/OWNERS](https://github.com/istio/istio/blob/master/pkg/OWNERS)~~ [nmittler]
- ~~[tests/integration/OWNERS](https://github.com/istio/istio/blob/master/tests/integration/OWNERS)~~ [nmittler]

Approvers can indicate their approval by writing `/approve` in a comment
Approvers can cancel approval by writing `/approve cancel` in a comment
</details>
<!-- META={"approvers":[]} -->

istio-testing commented:
@nmittler: The following tests **failed**, say `/retest` to rerun them all:

Test name | Commit | Details | Rerun command
--- | --- | --- | ---
prow/istio-integ-k8s-tests.sh | 4028704c6c49956f2e78897b520ba25b4e52e4cd | [link](https://k8s-gubernator.appspot.com/build/istio-prow/pr-logs/pull/istio_istio/13115/istio-integ-k8s-tests/6445) | `/test istio-integ-k8s-tests`
prow/istio-pilot-multicluster-e2e.sh | 56b17de09b780dd388cf141173d56624af7c2609 | [link](https://k8s-gubernator.appspot.com/build/istio-prow/pr-logs/pull/istio_istio/13115/istio-pilot-multicluster-e2e/10826) | `/test istio-pilot-multicluster-e2e`



<details>

Instructions for interacting with me using PR comments are available [here](https://git.k8s.io/community/contributors/guide/pull-requests.md).  If you have questions or suggestions related to my behavior, please file an issue against the [kubernetes/test-infra](https://github.com/kubernetes/test-infra/issues/new?title=Prow%20issue:) repository. I understand the commands that are listed [here](https://go.k8s.io/bot-commands).
</details>
<!-- test report -->

istio-testing commented:
@nmittler: The following tests **failed**, say `/retest` to rerun them all:

Test name | Commit | Details | Rerun command
--- | --- | --- | ---
prow/istio-pilot-multicluster-e2e.sh | 56b17de09b780dd388cf141173d56624af7c2609 | [link](https://k8s-gubernator.appspot.com/build/istio-prow/pr-logs/pull/istio_istio/13115/istio-pilot-multicluster-e2e/10826) | `/test istio-pilot-multicluster-e2e`
prow/istio-integ-k8s-tests.sh | 56b17de09b780dd388cf141173d56624af7c2609 | [link](https://k8s-gubernator.appspot.com/build/istio-prow/pr-logs/pull/istio_istio/13115/istio-integ-k8s-tests/6546) | `/test istio-integ-k8s-tests`



<details>

Instructions for interacting with me using PR comments are available [here](https://git.k8s.io/community/contributors/guide/pull-requests.md).  If you have questions or suggestions related to my behavior, please file an issue against the [kubernetes/test-infra](https://github.com/kubernetes/test-infra/issues/new?title=Prow%20issue:) repository. I understand the commands that are listed [here](https://go.k8s.io/bot-commands).
</details>
<!-- test report -->

istio-testing commented:
@nmittler: The following test **failed**, say `/retest` to rerun them all:

Test name | Commit | Details | Rerun command
--- | --- | --- | ---
prow/istio-integ-k8s-tests.sh | 05a0b6098d99c8f0101b74a6ce2049ca21647ebb | [link](https://k8s-gubernator.appspot.com/build/istio-prow/pr-logs/pull/istio_istio/13115/istio-integ-k8s-tests/6737) | `/test istio-integ-k8s-tests`



<details>

Instructions for interacting with me using PR comments are available [here](https://git.k8s.io/community/contributors/guide/pull-requests.md).  If you have questions or suggestions related to my behavior, please file an issue against the [kubernetes/test-infra](https://github.com/kubernetes/test-infra/issues/new?title=Prow%20issue:) repository. I understand the commands that are listed [here](https://go.k8s.io/bot-commands).
</details>
<!-- test report -->

nmittler commented:
@howardjohn I had to rebase ... PTAL

howardjohn commented:
/lgtm
Remove the hold label if you want this in

istio-testing commented:
[APPROVALNOTIFIER] This PR is **APPROVED**

This pull-request has been approved by: *<a href="https://github.com/istio/istio/pull/13115#issuecomment-482813613" title="LGTM">howardjohn</a>*, *<a href="https://github.com/istio/istio/pull/13115#" title="Author self-approved">nmittler</a>*

The full list of commands accepted by this bot can be found [here](https://go.k8s.io/bot-commands).

The pull request process is described [here](https://git.k8s.io/community/contributors/guide/owners.md#the-code-review-process)

<details >
Needs approval from an approver in each of these files:

- ~~[pkg/OWNERS](https://github.com/istio/istio/blob/master/pkg/OWNERS)~~ [nmittler]
- ~~[tests/integration/OWNERS](https://github.com/istio/istio/blob/master/tests/integration/OWNERS)~~ [nmittler]

Approvers can indicate their approval by writing `/approve` in a comment
Approvers can cancel approval by writing `/approve cancel` in a comment
</details>
<!-- META={"approvers":[]} -->

istio-testing commented:
@nmittler: The following test **failed**, say `/retest` to rerun them all:

Test name | Commit | Details | Rerun command
--- | --- | --- | ---
prow/istio-integ-k8s-tests.sh | 05a0b6098d99c8f0101b74a6ce2049ca21647ebb | [link](https://k8s-gubernator.appspot.com/build/istio-prow/pr-logs/pull/istio_istio/13115/istio-integ-k8s-tests/6791) | `/test istio-integ-k8s-tests`



<details>

Instructions for interacting with me using PR comments are available [here](https://git.k8s.io/community/contributors/guide/pull-requests.md).  If you have questions or suggestions related to my behavior, please file an issue against the [kubernetes/test-infra](https://github.com/kubernetes/test-infra/issues/new?title=Prow%20issue:) repository. I understand the commands that are listed [here](https://go.k8s.io/bot-commands).
</details>
<!-- test report -->
